lib/transaction.py

```diff
@@ -689,20 +689,48 @@ def nHashType(cls):
         warnings.warn("warning: deprecated tx.nHashType()", FutureWarning, stacklevel=2)
         return 0x01 | (cls.SIGHASH_FORKID + (cls.FORKID << 8))
 
-    def serialize_preimage(self, i, nHashType=0x00000041):
+    def calc_common_sighash(self, use_cache=False):
+        """ Calculate the common sighash components that are used by
+        transaction signatures. If `use_cache` enabled then this will return
+        already-computed values from the `._cached_sighash` attribute, or
+        compute them if necessary (and then store).
+
+        For transactions with N inputs and M outputs, calculating all sighashes
+        takes only O(N + M) with the cache, as opposed to O(N^2 + NM) without
+        the cache.
+
+        Returns three 32-long bytes objects: (hashPrevouts, hashSequence, hashOutputs).
+
+        Warning: If you modify non-signature parts of the transaction
+        afterwards, this cache will be wrong! """
+        try:
+            if use_cache:
+                return self._cached_sighash
+        except AttributeError:
+            pass
+
+        inputs = self.inputs()
+        outputs = self.outputs()
+        hashPrevouts = Hash(bfh(''.join(self.serialize_outpoint(txin) for txin in inputs)))
+        hashSequence = Hash(bfh(''.join(int_to_hex(txin.get('sequence', 0xffffffff - 1), 4) for txin in inputs)))
+        hashOutputs = Hash(bfh(''.join(self.serialize_output(o) for o in outputs)))
+
+        res = hashPrevouts, hashSequence, hashOutputs
+        if use_cache:
+            self._cached_sighash = (hashPrevouts, hashSequence, hashOutputs)
```

Comment:
why not `self._cached_sighash` = res?

This way you benefit from the typo-safety of re-using that same data (as opposed to actual **type-safety**, which python lacks).